src/OpenTelemetry.Instrumentation.Runtime/RuntimeMetrics.cs

```diff
@@ -76,28 +76,51 @@ static RuntimeMetrics()
                 unit: "bytes",
                 description: "The amount of committed virtual memory for the managed GC heap, as observed during the latest garbage collection. Committed virtual memory may be larger than the heap size because it includes both memory for storing existing objects (the heap size) and some extra memory that is ready to handle newly allocated objects in the future. The value will be unavailable until at least one garbage collection has occurred.");
 
-            // TODO: change to ObservableUpDownCounter
-            MeterInstance.CreateObservableGauge(
-                $"{metricPrefix}gc.heap.size",
-                () =>
+            // GC.GetGCMemoryInfo().GenerationInfo[i].SizeAfterBytes is better but it has a bug in .NET 6. See context in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/496
+            Func<int, ulong> getGenerationSize = null;
+            bool isCodeRunningOnBuggyRuntimeVersion = Environment.Version.Major == 6;
+            if (isCodeRunningOnBuggyRuntimeVersion)
+            {
+                MethodInfo mi = typeof(GC).GetMethod("GetGenerationSize", BindingFlags.NonPublic | BindingFlags.Static);
+                if (mi != null)
                 {
-                    if (!IsGcInfoAvailable)
-                    {
-                        return Array.Empty<Measurement<long>>();
-                    }
+                    getGenerationSize = mi.CreateDelegate<Func<int, ulong>>();
+                }
+            }
 
-                    var generationInfo = GC.GetGCMemoryInfo().GenerationInfo;
-                    Measurement<long>[] measurements = new Measurement<long>[generationInfo.Length];
-                    int maxSupportedLength = Math.Min(generationInfo.Length, GenNames.Length);
-                    for (int i = 0; i < maxSupportedLength; ++i)
+            // Either Environment.Version is not 6 or (it's 6 but internal API GC.GetGenerationSize is valid)
+            if (!isCodeRunningOnBuggyRuntimeVersion || getGenerationSize != null)
+            {
+                // TODO: change to ObservableUpDownCounter
+                MeterInstance.CreateObservableGauge(
+                    $"{metricPrefix}gc.heap.size",
```

Comment:
I looked at `RuntimeMetricsTests.cs` and I have a question it would not be good to test the exported metrics

Reply:
I didn't quite understand the sentence. Are you suggesting to add unit tests? I think it'd be a good idea since a lot were deleted in the test file but none was added. I can do it in a follow up PR.

Reply:
Added in here #510.